File: src/translated/readdata.cpp
//
//  readdata.cpp
//
//  Created by Eliana Fernandes
//  Copyright (c) 2015 Eliana Fernandes. All rights reserved.
//
#include <iostream>
#include <fstream> // file handling
#include <math.h> // sqrt

#include "readdata.h"

using namespace std;

///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// //////////
///////// ///////// ///////// /////////              FUNCTIONS        ///////// ///////// ///////// /////////
data readData(const char *fichIn){
    //cout.precision(3);
    data dados;
    double aux;
    ///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// /////////
    // read input from file
    ///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// /////////
    ifstream inFile(fichIn);
    // Check input file
    if ( !inFile ) {
        cerr<<"Input file could not be opened\n";
        system ("pause");
	}
    
    // Read number of nodes and number of products
	inFile >>  dados.nbNodes >> dados.nbProducts;
    point *coordenadas;
    
    int i,j,p;
    
	// Memory allocation
    coordenadas = new (nothrow) point[dados.nbNodes];
    if (coordenadas == 0)
        cout<<"Error: memory could not be allocated\n";
    
    dados.w = new (nothrow) double** [dados.nbNodes];
    if (dados.w == 0)
        cout<<"Error: memory could not be allocated\n";
    for (i=0; i<dados.nbNodes; ++i) {
        dados.w[i]= new (nothrow) double* [dados.nbNodes];
        if (dados.w [i] == 0)
            cout<<"Error: memory could not be allocated\n";
        for (j=0; j<dados.nbNodes; ++j){
            dados.w[i][j]= new (nothrow) double [dados.nbProducts];
            if (dados.w[i][j] == 0)
                cout<<"Error: memory could not be allocated\n";
        }
        
    }
    dados.chi = new (nothrow) double [dados.nbProducts];
    if (dados.chi == 0)
        cout<<"Error: memory could not be allocated\n";
    dados.alpha = new (nothrow) double [dados.nbProducts];
    if (dados.alpha == 0)
        cout<<"Error: memory could not be allocated\n";
    dados.delta = new (nothrow) double [dados.nbProducts];
    if (dados.delta == 0)
        cout<<"Error: memory could not be allocated\n";
    
    dados.d = new (nothrow) double* [dados.nbNodes];
    if (dados.d == 0)
        cout<<"Error: memory could not be allocated\n";
    for (i=0; i<dados.nbNodes; ++i){
        dados.d[i]= new (nothrow) double [dados.nbNodes];
        if (dados.d[i] == 0)
            cout<<"Error: memory could not be allocated\n";
    }
    
    dados.g = new (nothrow) double [dados.nbNodes];
    if (dados.g == 0)
        cout<<"Error: memory could not be allocated\n";
    
    dados.L = new (nothrow) int [dados.nbNodes];
    if (dados.L == 0)
        cout<<"Error: memory could not be allocated\n";
    
    dados.f = new (nothrow) double * [dados.nbNodes];
    if (dados.f == 0)
        cout<<"Error: memory could not be allocated\n";
    for (i=0; i<dados.nbNodes; ++i){
        dados.f[i] = new (nothrow) double [dados.nbProducts];
        if (dados.f[i] == 0)
            cout<<"Error: memory could not be allocated\n";
    }
    dados.Gamma = new (nothrow) double * [dados.nbNodes];
    if (dados.Gamma == 0)
        cout<<"Error: memory could not be allocated\n";
    for (i=0; i<dados.nbNodes; ++i){
        dados.Gamma[i] = new (nothrow) double [dados.nbProducts];
        if (dados.Gamma [i] == 0)
            cout<<"Error: memory could not be allocated\n";
    }
    dados.O = new (nothrow) double * [dados.nbNodes];
    if (dados.O == 0)
        cout<<"Error: memory could not be allocated\n";
    for (i=0; i<dados.nbNodes; ++i){
        dados.O[i] = new (nothrow) double [dados.nbProducts];
        if (dados.O[i] == 0)
            cout<<"Error: memory could not be allocated\n";
    }
    dados.D = new (nothrow) double * [dados.nbNodes];
    if (dados.D == 0)
        cout<<"Error: memory could not be allocated\n";
    for (i=0; i<dados.nbNodes; ++i){
        dados.D[i] = new (nothrow) double [dados.nbProducts];
        if (dados.D[i] == 0)
            cout<<"Error: memory could not be allocated\n";
    }
    
    // Read coordinates
	for (i=0; i<dados.nbNodes; i++)
        inFile >> coordenadas[i].x >> coordenadas[i].y;
    
	// Calculate distances
	for (i=0; i< dados.nbNodes; i++)
		for (int j=0; j< dados.nbNodes; j++) {
			aux=(coordenadas[i].x-coordenadas[j].x)*(coordenadas[i].x-coordenadas[j].x);
            aux+=(coordenadas[i].y-coordenadas[j].y)*(coordenadas[i].y-coordenadas[j].y);
			dados.d[i][j]=sqrt(aux)/1000.0;
		}
    
	// Read flows
	for (p=0; p<dados.nbProducts; ++p)
		for (i=0; i< dados.nbNodes; ++i)
			for (j=0; j< dados.nbNodes; ++j)
				inFile >> dados.w[i][j][p];
    
	// Read collection, discount and distribution costs
	for (p=0; p<dados.nbProducts; ++p)
		inFile >> dados.chi[p];
	for (p=0; p<dados.nbProducts; ++p)
		inFile >> dados.alpha[p];
	for (p=0; p<dados.nbProducts; ++p)
		inFile >> dados.delta[p];
    
	// Read set-up costs
	for (i=0; i< dados.nbNodes; i++)
		inFile >> dados.g[i];
    
	// Read maximum numbder of products that can be handled in each node
	for (i=0; i< dados.nbNodes; i++)
		inFile >> dados.L[i];
    
	// Read capacities
	for (i=0; i< dados.nbNodes; i++)
		for (p=0; p<dados.nbProducts; ++p){
			inFile >> dados.Gamma[i][p];

#ifdef DAT_HIST
			ofstream myfile;
			myfile.open ("history1.txt",ios::app);
			myfile << "dados.Gamma["<<i<<"]["<<p<<"]= " << dados.Gamma[i][p] << endl;
			myfile.close();
#endif
		}
    // Read set up costs per product
	for (i=0; i< dados.nbNodes; i++)
		for (p=0; p<dados.nbProducts; ++p)
			inFile >> dados.f[i][p];
    
    // Update the g[i] values according with the paper
	for (i=0; i<dados.nbNodes; i++){
		dados.g[i]=0.0;
		for (p=0; p<dados.nbProducts; ++p)
			dados.g[i] = dados.g[i] + dados.f[i][p];
		dados.g[i]= 2*dados.g[i]/(dados.nbProducts*1.0);
	}
    
    // Close file for reading
	inFile.close();
    
    // Calculate total flows originated at the nodes
	for (i=0; i<dados.nbNodes; ++i)
		for (p=0; p<dados.nbProducts; ++p) {
			dados.O[i][p]=0.0;
			for (j=0; j<dados.nbNodes; ++j)
				dados.O[i][p]+=dados.w[i][j][p];
#ifdef DAT_HIST
			ofstream myfile;
			myfile.open ("history1.txt",ios::app);
			myfile << "dados.O["<<i<<"]["<<p<<"]= " << dados.O[i][p] << endl;
			myfile.close();
#endif
        }
    
	// Calculate total flows destined to the nodes
	for (i=0; i<dados.nbNodes; ++i)
		for (p=0; p<dados.nbProducts; ++p) {
			dados.D[i][p]=0.0;
			for (j=0; j<dados.nbNodes; ++j)
				dados.D[i][p]+=dados.w[j][i][p];
        }
    
    
    delete [] coordenadas;
    return (dados);
}
///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// ///////// //////////

File: src/antcolony/ReadData.java
package antcolony;

/**
 * Readdata.java
 * Translated from readdata.cpp + readdata.h
 * Original by Eliana Fernandes
 * Copyright (c) 2015 Eliana Fernandes. All rights reserved.
 */
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import antcolony.constants.AcoVar;

public class ReadData {
	
	private static final Logger log = LogManager.getLogger(ReadData.class);

    /**
     * Data structure holding the problem instance
     */
    public static class Data {
    	
        public int nbNodes = 0;
        public int nbProducts = 0;
        public double[][][] w;  // w[p][i][j] = flow from i to j for product p
        public double[][] d;    // d[i][j] = distance between i and j
        public double[] g;      // g[i] = fixed cost to open hub at i
        public double[][] f;    // f[p][i] = fixed cost to dedicate hub i to product p
        public double[][] gamma;// Gamma[p][i] = capacity of hub i for product p
        public int[] L;         // L[i] = max products hub i can handle
        public double[] chi;    // chi[p] = collection cost factor for p
        public double[] delta;  // delta[p] = distribution cost factor for p
        public double[] alpha;  // alpha[p] = inter-hub transfer cost factor for p
        public double[][] originatedFlow;    // O[p][i] = total outflow from i for p
        public double[][] destinedFlow;    // D[p][i] = total inflow to i for p

        public Data(int nbNodes, int nbProducts) {
            this.nbNodes = nbNodes;
            this.nbProducts = nbProducts;

            w = new double[nbProducts][nbNodes][nbNodes];
            d = new double[nbNodes][nbNodes];
            g = new double[nbNodes];
            f = new double[nbProducts][nbNodes];
            gamma = new double[nbProducts][nbNodes];
            L = new int[nbNodes];
            chi = new double[nbProducts];
            delta = new double[nbProducts];
            alpha = new double[nbProducts];
            originatedFlow = new double[nbProducts][nbNodes];
            destinedFlow = new double[nbProducts][nbNodes];
        }
    }

    /**
     * Temporary point for coordinates
     */
    private static class Point {
        double x;
        double y;
    }

    /**
     * Reads the problem instance from file
     */
    public static Data readData(String fichIn) {
        Data dados = null;
        String input = AcoVar.INPUT_PATH + fichIn;
        try (Scanner scanner = new Scanner(new File(input))) {

            // Read nbNodes and nbProducts
            int nbNodes = scanner.nextInt();
            int nbProducts = scanner.nextInt();

            dados = new Data(nbNodes, nbProducts);

            // Read coordinates and compute distances
            Point[] coordenadas = new Point[nbNodes];
            for (int i = 0; i < nbNodes; i++) {
                coordenadas[i] = new Point();
                coordenadas[i].x = scanner.nextDouble();
                coordenadas[i].y = scanner.nextDouble();
            }

            // Compute Euclidean distances
            for (int i = 0; i < nbNodes; i++) {
                for (int j = 0; j < nbNodes; j++) {
                    double dx = coordenadas[i].x - coordenadas[j].x;
                    double dy = coordenadas[i].y - coordenadas[j].y;
                    dados.d[i][j] = Math.sqrt(dx * dx + dy * dy);
                }
            }

            // Read flows w[p][i][j]
            for (int i = 0; i < nbNodes; i++) {
                for (int j = 0; j < nbNodes; j++) {
                    for (int p = 0; p < nbProducts; p++) {
                        dados.w[p][i][j] = scanner.nextDouble();
                    }
                }
            }

            // Read chi, alpha, delta
            for (int p = 0; p < nbProducts; p++) {
                dados.chi[p] = scanner.nextDouble();
            }
            for (int p = 0; p < nbProducts; p++) {
                dados.alpha[p] = scanner.nextDouble();
            }
            for (int p = 0; p < nbProducts; p++) {
                dados.delta[p] = scanner.nextDouble();
            }

            // Read hub opening costs g[i]
            for (int i = 0; i < nbNodes; i++) {
                dados.g[i] = scanner.nextDouble();
            }

            // Read max products per hub L[i]
            for (int i = 0; i < nbNodes; i++) {
                dados.L[i] = scanner.nextInt();
            }

            // Read capacities gamma[p][j]
            for (int i = 0; i < nbNodes; i++) {
                for (int p = 0; p < nbProducts; p++) {
                    dados.gamma[p][i] = scanner.nextDouble();
                    if (AcoVar.DAT_HIST) {
                    	log.info("dados.Gamma[{}][{}]= " +i, p,  dados.gamma[p][i]);
                    }
                }
            }

            // Read dedication costs f[p][i]
            for (int i = 0; i < nbNodes; i++) {
                for (int p = 0; p < nbProducts; p++) {
                    dados.f[p][i] = scanner.nextDouble();
                }
            }

            // Update g[i] as per paper: average 2 * sum f[p][i] / nbProducts
            for (int i = 0; i < nbNodes; i++) {
                double sumF = 0.0;
                for (int p = 0; p < nbProducts; p++) {
                    sumF += dados.f[p][i];
                }
                dados.g[i] = 2 * sumF / nbProducts;
            }

            // Compute total outflows O[i][p]
            for (int i = 0; i < nbNodes; i++) {
                for (int p = 0; p < nbProducts; p++) {
                    dados.originatedFlow[p][i] = 0.0;
                    for (int j = 0; j < nbNodes; j++) {
                        dados.originatedFlow[p][i] += dados.w[p][i][j];
                    }
                    if (AcoVar.DAT_HIST) {
                    	log.info("dados.O[{}][{}]= " ,i,p, dados.originatedFlow[p][i]);
                    }
                    
                }
            }

            // Compute total inflows D[i][p]
            for (int i = 0; i < nbNodes; i++) {
                for (int p = 0; p < nbProducts; p++) {
                    dados.destinedFlow[p][i] = 0.0;
                    for (int j = 0; j < nbNodes; j++) {
                        dados.destinedFlow[p][i] += dados.w[p][j][i];
                    }
                }
            }

        } catch (FileNotFoundException e) {
        	log.error("Input file could not be opened: {}", fichIn);
            System.exit(1);
        }

        return dados;
    }
}
